runtime/Cpp/runtime/src/atn/PredictionContext.cpp

```diff
@@ -389,11 +389,11 @@ std::string PredictionContext::toDOTString(const Ref<const PredictionContext> &c
   });
 
   for (const auto &current : nodes) {
-    if (is<const SingletonPredictionContext>(current)) {
+    if (current->getContextType() == PredictionContextType::SINGLETON || current->getContextType() == PredictionContextType::EMPTY) {
```

Comment:
Since we're calling the virtual `getContextType()` multiple times in this block (here and below, in total 3 times), it might be worthwhile calling it only once, remembering the result and do the cheap comparisons without the virtual method call overhead multiple times. Picking this as an example, there are also the other places that might need to call virtual functions multiple times.

```c++
  for (const auto &current : nodes) {
     const PredictionContextType contextType = current->getContextType();
      if (contextType == PredictionContextType::SINGLETON || contextType == PredictionContextType::EMPTY) {
            //...
            if (contextType == PredictionContextType::EMPTY) {
              // ...
            }
       }
  }
```

Reply:
Fixed.